File: js/sliding_window/3_longest_substring_without_repeating_characters.ts
function lengthOfLongestSubstring(s: string): number {
  if (!s.length) return 0;

  let left = 0;
  const set: Set<string> = new Set();
  set.add(s[left]);

  let longest: number = set.size;

  for (let right: number = left + 1; right < s.length; right++) {
    while (set.has(s[right])) {
      set.delete(s[left]);
      left += 1;
    }
    set.add(s[right]);

    longest = set.size > longest ? set.size : longest;
  }

  return longest;
}

(() => {
  // Test cases

  // result: 3
  console.log(lengthOfLongestSubstring('abcabcbb'));

  // result: 1
  console.log(lengthOfLongestSubstring('bbbbb'));

  // result: 3
  console.log(lengthOfLongestSubstring('pwwkew'));

  // result: 3
  console.log(lengthOfLongestSubstring('dvdf'));
})();

File: ruby/sliding_window/3_longest_substring_without_repeating_characters.rb
require "set"

# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  return s.size if s.size <= 1

  left = 0
  set = Set[s[left]]

  longest = set.size

  (left + 1..s.length - 1).each do |right|
    while set.include? s[right]
      set.delete s[left]
      left += 1
    end

    set.add(s[right])
    longest = set.size if set.size > longest
  end

  longest
end

def test
  # Test cases

  # result: 3
  puts length_of_longest_substring("abcabcbb")

  # result: 1
  puts length_of_longest_substring("bbbbb")

  # result: 3
  puts length_of_longest_substring("pwwkew")

  # result: 3
  puts length_of_longest_substring("dvdf")
end

test
